location tag and map implementation - In progress

lalithr95 opened the pull request:
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] tests pass -- `rake test`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request are descriptively named
* [x] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer
* [ ] `development.sqlite.example` has been updated if any database migrations were added

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

lalithr95 commented:
@jywarren There is an input field to add user location along with other tags. I added a functionality to fetch exact location from browser using browser navigator. This can be an option if user doesn't want to type location in text field. This would really reduce time for user to add location. What do you think ?
<img width="1279" alt="user map" src="https://cloud.githubusercontent.com/assets/4562611/16001248/ab28d560-316e-11e6-89ba-c4f73adc253b.png">

I will be working on user map where all the user will be displayed as markers in map. Which map would you suggest Google Map (or) Leaflet (similar to /places) ?

jywarren commented:
I like it -- also, though, if you can please use a Leaflet map instead of a google map, that would be excellent -- we use Leaflet elsewhere, as in http://publiclab.org/archive. But perhaps we'd need a different layer (we use a MapBox layer but there are others) to show streets and such better?

jywarren commented:
Ah, sorry, reading too fast -- i see you asked specifically about Leaflet vs. Google at the bottom of your post. Glad we're thinking on the same wavelength!

lalithr95 commented:
I will use Leaflet flavoured map with different layer. Will update this PR.

lalithr95 commented:
@jywarren I changed map in Location from Google map to leaflet map. Let me know your views on style specification (height, width)
<img width="1279" alt="leaflet" src="https://cloud.githubusercontent.com/assets/4562611/16069280/d2b18b0e-32ea-11e6-9c80-79b66486aeae.png">

jywarren commented:
Oh cool! How do we represent precision on the map, with a circle instead of
a point? Imagine this giving you feedback about how much precision you're
revealing in your location.
On Jun 15, 2016 1:48 AM, "Lalith Rallabhandi" <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren> I changed map in Location from
> Google map to leaflet map. Let me know your views on style specification
> (height, width)
> [image: leaflet]
> <https://cloud.githubusercontent.com/assets/4562611/16069280/d2b18b0e-32ea-11e6-9c80-79b66486aeae.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-226093988>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AABfJ6NgFyqUL_duJKWV9OIfhNyw1U32ks5qL5IvgaJpZM4Izmq->
> .
>

lalithr95 commented:
Currently for latitude and longitude I'm using 15 as precision which pretty good to identify exact location. 
I can implement a circle marker as well, How much radius would be good for circle, as point shows exact location ?

Thanks !

jywarren commented:
Maybe @ebarry can help you come up with an initial proposal and post some design ideas in a research note to get input from community members. I think it'll depend on people's thoughts on privacy. In the meantime, do you think we should merge this or keep working on it a bit in the feature branch?

lalithr95 commented:
@jywarren I would like to keep it as feature branch and will merge once location-tag for user has been implemented. Would like hear more thoughts :)

ebarry commented:
Hi Lalith,
Excited about this!

Have you ever used AirBNB?

Setting aside arguments on "enclosure technologies", AirBNB has figured out
how to protect user privacy while giving a sense of location. How? Their
circles are not centered around the exact lat/lng of the address. We need
something like this -- a generalized location around which to center a
bigger circle. Let's discuss if merely knocking off significant digits from
the precision on the lat/lng is enough to count as "generalized" or if we
need to do something else.

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>


On Fri, Jun 17, 2016 at 1:08 PM, Lalith Rallabhandi <
notifications@github.com> wrote:

> @jywarren <https://github.com/jywarren> I would like to keep it as
> feature branch and will merge once location-tag for user has been
> implemented. Would like hear more thoughts :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-226825968>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAJ2n7bzJLAE0LF4TI3aO6pV9w-eRL2aks5qMtRxgaJpZM4Izmq->
> .
>

lalithr95 commented:
@ebarry I added few commits which display similar circle in map. Below is the screenshot for the same. Also I'm storing other data such as country, state, city by parsing the location. These data can be used in next stages.
<img width="1279" alt="circl with map" src="https://cloud.githubusercontent.com/assets/4562611/16198043/058fbb00-3722-11e6-9928-ca59ec4000cd.png">

ebarry commented:
Hi Lalith this is great!
Can you further describe the circles, specifically what is used as their
centerpoint?

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>


On Mon, Jun 20, 2016 at 10:33 AM, Lalith Rallabhandi <
notifications@github.com> wrote:

> @ebarry <https://github.com/ebarry> I added few commits which display
> similar circle in map. Below is the screenshot for the same. Also I'm
> storing other data such as country, state, city by parsing the location.
> These data can be used in next stages.
> [image: circl with map]
> <https://cloud.githubusercontent.com/assets/4562611/16198043/058fbb00-3722-11e6-9928-ca59ec4000cd.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-227159976>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAJ2nxQzduzIHvwmFuv7jTyjmoh0Eyonks5qNqS8gaJpZM4Izmq->
> .
>

lalithr95 commented:
From the leaflet circle API I'm using circle with radius as 300 and center as lat, long of the user. By which user privacy should not be disturbed. I have added 2 ways to add location to user.

1. Through browser navigator api (available in most browsers). If user clicks on map button and allows permission to access user browser location on publiclab, it would fetch location (We get exact location of user, similar to way you point current location in uber app) and plot a circle in the map.

2. User can type location manually and the input field is autocomplete. Once user selects a location from dropdown list (location may not be accurate), app saves location, loads map and plots a circle in the map.

jywarren commented:
Perhaps there needs to be a "don't share exact location" checkbox, which
when checked, auto-truncates and/or adds randomness to the location (so the
center of the circle is not actually the location) and shows the circle.
This makes privacy more explicit and easier to understand.

* When unchecking the box, you'd then see a more specific point.
* When checked, we should show (perhaps in grey, below the location input
field) the exact data we're storing in the database -- i.e. including how
it is obscured/truncated/etc. -- perhaps with a notice like "exact location
obscured"
* The checkbox could be checked by default

This way, there's also no ambiguity about what exact location means, and
that there's no difference between what's being shown and what's actually
stored.



On Mon, Jun 20, 2016 at 4:51 PM, Lalith Rallabhandi <
notifications@github.com> wrote:

> From the leaflet circle API I'm using circle with radius as 300 and center
> as lat, long of the user. By which user privacy should not be disturbed. I
> have added 2 ways to add location to user.
>
>    1.
>
>    Through browser navigator api (available in most browsers). If user
>    clicks on map button and allows permission to access user browser location
>    on publiclab, it would fetch location (We get exact location of user,
>    similar to way you point current location in uber app) and plot a circle in
>    the map.
>    2.
>
>    User can type location manually and the input field is autocomplete.
>    Once user selects a location from dropdown list (location may not be
>    accurate), app saves location, loads map and plots a circle in the map.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-227165595>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AABfJw93JSReUgipd_IlM6rx1PmCvtcAks5qNqkEgaJpZM4Izmq->
> .
>

lalithr95 commented:
@jywarren I implemented a checkbox to display exact location or truncated location. For exact location I'm using a marker and for truncated location I will be using circle to point approximate location.

We can merge this once reviewed !!

jywarren commented:
Hi, Lalith - can you post a screenshot or two so @ebarry can take a look and provide feedback, especially as I'm not as available this week?

lalithr95 commented:
If user has a location added in his profile, user profile displays location (state, country)
<img width="1279" alt="location" src="https://cloud.githubusercontent.com/assets/4562611/16260825/ad86f3d6-3886-11e6-8fbe-6d723eb9c187.png">

There is a checkbox location privacy which on checked, displays user exact location. If user doesn't want to share location a circle will be plotted which hides exact location of user.  I would like to add a message below checkbox so that it clears the usage of privacy for user. @ebarry  looking forward for your feedback.
<img width="1279" alt="display exact location" src="https://cloud.githubusercontent.com/assets/4562611/16260834/b5c8943c-3886-11e6-89c2-aa4053fee7f5.png">
<img width="1279" alt="displays approx location" src="https://cloud.githubusercontent.com/assets/4562611/16260835/b5f9da2e-3886-11e6-8dde-4175f0511d93.png">

lalithr95 commented:
Display Users in the map based on their location privacy
<img width="1279" alt="user display" src="https://cloud.githubusercontent.com/assets/4562611/16529471/1dd6fbb6-3f91-11e6-83bc-eef11d842dd0.png">

User name pop-up
<img width="1279" alt="user name popup" src="https://cloud.githubusercontent.com/assets/4562611/16529490/38b106c0-3f91-11e6-862b-118f032d2457.png">

Display users based on skill as search filter
<img width="1279" alt="display users based on skill" src="https://cloud.githubusercontent.com/assets/4562611/16529497/43154c7a-3f91-11e6-955e-214fdfba67c9.png">

ebarry commented:
Looking great Lalith! You have done such incredible work and covered so much ground that i think it might be time for the Rich Profiles team to get on a googleHangout next week and talk through it all. What time zone are you in? We are in UTC-5 and UTC-6

lalithr95 commented:
Thanks a lot @ebarry, I'm in UTC-5 timezone

ebarry commented:
what? Where are you actually? There are tons of Public Labbers all over
UTC-5

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>


On Fri, Jul 1, 2016 at 1:46 PM, Lalith Rallabhandi <notifications@github.com
> wrote:

> Thanks a lot @ebarry <https://github.com/ebarry>, I'm in UTC-5 timezone
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-230007809>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAJ2n-pwoh4IiocpXRk2bX09JLYk3pkQks5qRVKCgaJpZM4Izmq->
> .
>

lalithr95 commented:
I'm in Ottawa, Canada. Really excited to meet them. Where do you stay ?

ebarry commented:
i'm in Brooklyn, there are a lot of people in NYC, also a lot in Boston and
the northeast US in general. Here's stuff tagged Canada
https://publiclab.org/tag/canada (mostly Quebec and BC/Alberta) but also
there was a high school student in Ottawa working with spectroscopy who
wanted to get local Public Labbers together
https://publiclab.org/profile/Haripriya.

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>


On Fri, Jul 1, 2016 at 1:58 PM, Lalith Rallabhandi <notifications@github.com
> wrote:

> I'm in Ottawa, Canada. Really excited to meet them. Where do you stay ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-230010447>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAJ2n3NgyL1c2jmawScjcsHTgRyHvmKFks5qRVVRgaJpZM4Izmq->
> .
>

lalithr95 commented:
Awesome !! I will try to connect with people here.

lalithr95 commented:
@ebarry Looks like User follow relationship is not implemented. I'm looking into it, so that publiclab user can follow other user.

jywarren commented:
If I'm following this, we're not quite ready yet? The discussion on hexbins is still moving? Although if you like, I could pull this in (assuming it's well tested etc) as you continue refining location systems, so it doesn't drift too far from the master branch.

lalithr95 commented:
Yup this is update. If we can merge this branch I can add changes on hexbin on top of it. Yeah exactly all the test are currently passing now 👍

jywarren commented on app/views/layouts/application.html.erb:
```diff
@@ -32,6 +32,9 @@
     <%= javascript_include_tag "application" %>
     <%= csrf_meta_tags %>
 
+    <script src="https://maps.googleapis.com/maps/api/js?libraries=places&language=en&key=AIzaSyDWgc7p4WWFsO3y0MTe50vF4l4NUPcPuwE"></script>
```
Do we need this for every page? I think perhaps it should be in only more specific templates.

lalithr95 replied:
Yeah I'll include in specific templates.

jywarren commented on app/views/layouts/_header.html.erb:
```diff
@@ -49,6 +49,7 @@
             <li><a href="/wiki/"><%= t('layout._header.projects.research.publiclab_wiki') %></a></li>
             <li><a href="/tools"><%= t('layout._header.projects.research.tools') %></a></li>
             <li><a href="/places"><%= t('layout._header.projects.research.places') %></a></li>
+            <li><a href="/users/map"><%= t('layout._header.projects.research.users_map') %></a></li>
```
Here, the link would be `/users/map` -- @ebarry - can you suggest a better URL for a map of members? maybe `/people/map` or `/map/people`?

jywarren commented on app/models/user.rb:
```diff
@@ -8,7 +8,7 @@ def validate(record)
 class User < ActiveRecord::Base
   self.table_name = 'rusers'
-  attr_accessible :username, :email, :password, :password_confirmation, :openid_identifier, :key, :photo, :photo_file_name
+  attr_accessible :username, :email, :password, :password_confirmation, :openid_identifier, :key, :photo, :photo_file_name, :location_privacy
```
Is there a reason to have location privacy be set in a unique column on user, rather than on the location table? Could you imagine someone wanting one location private and another not, both associated with them in some way? Perhaps this belongs on the location tag table.

jywarren commented on app/models/location_tag.rb:
```diff
@@ -0,0 +1,4 @@
+class LocationTag < ActiveRecord::Base
+  attr_accessible :lat, :location, :long, :uid, :country, :state, :city
```
`lon` or `lng` (I prefer `lon`) is a bit more standard than `long` -- would you mind changing this parameter name?

jywarren commented on app/models/drupal_users.rb:
```diff
@@ -11,6 +11,7 @@ class DrupalUsers < ActiveRecord::Base
   has_many :answers, :foreign_key => :uid
   has_many :answer_selections, :foreign_key => :user_id
   has_many :drupal_comments, :foreign_key => 'uid'
+  has_one :location_tag, :foreign_key => 'uid', :dependent => :destroy
```
Here, I thought we had decided that location tags would be associated with a profile tag, not with a person. But if there's a strong advantage to a `has_one` here, I'm open to it. 

If we do associate with a user record, we should use `user` and not `drupal_user`, which we're trying to deprecate.

jywarren commented on config/locales/views/layouts/_header/en.yml:
```diff
@@ -14,6 +14,7 @@ en:
           tools: "Tools"
           places: "Places"
           kits: "Kits"
+          users_map: "Users map"
```
Again here, I think we'd like input from @ebarry on naming this something other than "Users map".

jywarren commented on config/routes.rb:
```diff
@@ -159,8 +159,12 @@
   match 'info/tags/create/:id' => 'user_tags#create'
   match 'info/tags/delete/:id' => 'user_tags#delete'
+  match 'info/location/create/:id' => 'location_tags#create'
+  match 'info/privacy' => 'users#privacy'
```
I'm curious - what do you think of paths like 'profile/suggested...', 'profile/tags/create/:id' etc -- instead of "info"?

jywarren commented:
Hi, Lalith - sorry, I know this is a lot of additional changes, but as this is a big feature and an important one, I want to get this right before merging (possibly Friday). I would deeply appreciate if you could look through these questions, suggestions, and requests, and rebase your changes on top of master again. 

There has also been a change to the database since I merged in #598, so we may need to ensure the timestamps of the migration are in order. 

Thank you /very much/ for your hard work! This is exciting to see.

lalithr95 commented:
@jywarren Thanks for the review. I will fix the issues and should be able to merge asap.

jywarren commented:
Hi, Lalith, any update? I'll be pulling things in today if ready. Please also remember to rebase if you push a new commit. Thanks!

lalithr95 commented:
@jywarren rebased with master and updated few changes. I'm looking for path content suggestion so that I can update the links in routes as well as views. I used long as the User model contains column named long. I can change name ? All tests are passing. I added a new follower and following to page to this branch, thats the reason took time. I will create a separate PR for user relationship in the profile. 

Thanks !

jywarren commented:
Hi, Lalith - by "looking for path content suggestion" do you mean whether to make routes be `/profile/x` vs. `/info/x`? I think `/profile/x` is probably best. 

I think `lon` is best, even though you're right that User uses `long` -- and feel free to change that in a migration, but please double check that it's not used anywhere else.

Is follower/following a listing of people following this person, and people this person is following? That's a great addition! But let's get this one finished and merged in on Friday and try for smaller self-contained PRs in the future. Once this is ready, please rebase once more and using the `--interactive` flag, can you squash some of these commits so that the changes are easier to follow in your commit log? 

Thanks for your hard work, Lalith!

lalithr95 commented:
Hi Jeff, I updated path location also with respective functional and unit tests. All tests are passing. Also added a new migration for column change. 

I squashed some of recent commits. Previous commit contains exact additions for that commit. It can be used in future reference. Do you want them to squash as well ?

Thanks for the review !

ebarry commented:
Thanks for all this amazing work Lalith -- this is a huge influence and
really a whole new phase of geographic organizing that you are bringing to
Public Lab community.

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>


On Thu, Jul 14, 2016 at 11:10 PM, Lalith Rallabhandi <
notifications@github.com> wrote:

> Hi Jeff, I updated path location also with respective functional and unit
> tests. All tests are passing. Also added a new migration for column change.
>
> I squashed some of recent commits. Previous commit contains exact
> additions for that commit. It can be used in future reference. Do you want
> them to squash as well ?
>
> Thanks for the review !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/578#issuecomment-232849677>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJ2n36HRw417JesAumECpjlcyxrHZJkks5qVvovgaJpZM4Izmq->
> .
>

jywarren commented:
This looks good; i'll test and merge today. In general, it's great for each commit to have a self-contained amount of content, so ones like "fix conflicts" could be squashed into whatever it fixed conflicts *for*. 

Is development.sqlite.example updated? Thanks!

jywarren commented:
Ah, i see development.sqlite.example is not updated. Can you add this last change and I'll merge? Thanks!

lalithr95 commented:
Updated development.sqlite.example Heading on merge 👍 
Thanks !

jywarren commented:
Testing!

jywarren commented:
All tests pass. Merging!

lalithr95 commented:
Thanks a lot for the review @jywarren I'm working on extension with a follow up PR.
